test/mpi/test_mpi_cuda.c

```diff
@@ -0,0 +1,695 @@
+    retain_and_push_primary_context(cu_dev_other);
+
+    mpi_pingpong(params->rank, alloc_mem_send.ptr, alloc_mem_recv.ptr,
+                 params->size);
+
+    pop_and_release_primary_context(cu_dev_other);
+
+    res = check_and_free_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    pop_and_release_primary_context(params->cu_dev);
+
+    return res;
+}
+
+static int check_allocator(const allocator_t *allocator, CUdevice cu_dev)
+{
+    alloc_mem_t alloc_mem;
+    int ret;
+
+    retain_and_push_primary_context(cu_dev);
+    alloc_mem = allocator->alloc(1);
+    if (alloc_mem.ptr == 0) {
+        ret = 0;
+        goto pop_and_release;
+    }
+
+    allocator->free(&alloc_mem);
+    ret = 1;
+
+pop_and_release:
+    pop_and_release_primary_context(cu_dev);
+    return ret;
+}
+
+static void run_test_for_all_sizes(test_params_t *params, int test_idx)
+{
+    int size_idx;
+    int res;
+
+    for (size_idx = 0; size_idx < ucs_static_array_size(sizes); ++size_idx) {
+        total_ucx_errors_and_warnings = 0;
+
+        params->size = sizes[size_idx];
+        res          = tests[test_idx].func(params);
+
+        res += total_ucx_errors_and_warnings;
+
+        if (params->rank == 0) {
+            MPI_Reduce(MPI_IN_PLACE, &res, 1, MPI_INT, MPI_SUM, 0,
+                       MPI_COMM_WORLD);
+        } else {
+            MPI_Reduce(&res, NULL, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);
+        }
+
+        PRINT_ROOT(TEST_NAME_SIZE_FMT ": %s\n", test_idx,
+                   params->allocator_send->name, params->allocator_recv->name,
+                   params->size, (res ? "FAIL" : "PASS"));
+    }
+}
+
+int main(int argc, char **argv)
+{
+#if ENABLE_MT
+    int required = MPI_THREAD_MULTIPLE;
+#else
+    int required = MPI_THREAD_SINGLE;
+#endif
+    int provided;
+    int comm_size;
+    int dev_count, dev_idx;
+    CUdevice cu_dev;
+    CUcontext cu_ctx;
+    int test_idx, allocator_send_idx, allocator_recv_idx;
+    test_params_t test_params;
+    const test_t *test;
+
+    MPI_Init_thread(&argc, &argv, required, &provided);
+    MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
+    if (comm_size != MPI_COMM_SIZE) {
+        PRINT_ROOT("This test requires %d processes, not %d\n", MPI_COMM_SIZE,
+                   comm_size);
+        goto out;
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &test_params.rank);
+
+    CUDA_CHECK(cuInit(0));
+    CUDA_CHECK(cuDeviceGetCount(&dev_count));
+    if (dev_count < MIN_DEV_COUNT) {
+        PRINT_ROOT("This test requires at least %d GPU\n", MIN_DEV_COUNT);
+        goto out;
+    }
+
+    for (dev_idx = dev_count - 1; dev_idx > -1; --dev_idx) {
+        CUDA_CHECK(cuDeviceGet(&cu_dev, dev_idx));
+        CUDA_CHECK(cuDevicePrimaryCtxRetain(&cu_ctx, cu_dev));
+    }
+
+    if ((test_params.rank == 1) && (dev_count > 1)) {
+        CUDA_CHECK(cuDeviceGet(&cu_dev, 1));
+    }
+
+    test_params.gold_data = create_gold_data(SIZE_L);
```

Comment:
can gold_data be just a global ptr, so there will be no need to pass it in params down till check func?

Reply:
made it a global var.